doc/file.man
.\" $File: file.man,v 1.133 2018/10/01 18:51:39 christos Exp $
.Dd October 1, 2018
.Dt FILE __CSECTION__
.Os
.Sh NAME
.Nm file
.Nd determine file type
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl bcdEhiklLNnprsSvzZ0
.Op Fl Fl apple
.Op Fl Fl extension
.Op Fl Fl mime-encoding
.Op Fl Fl mime-type
.Op Fl e Ar testname
.Op Fl F Ar separator
.Op Fl f Ar namefile
.Op Fl m Ar magicfiles
.Op Fl P Ar name=value
.Ar
.Ek
.Nm
.Fl C
.Op Fl m Ar magicfiles
.Nm
.Op Fl Fl help
.Sh DESCRIPTION
This manual page documents version __VERSION__ of the
.Nm
command.
.Pp
.Nm
tests each argument in an attempt to classify it.
There are three sets of tests, performed in this order:
filesystem tests, magic tests, and language tests.
The
.Em first
test that succeeds causes the file type to be printed.
.Pp
The type printed will usually contain one of the words
.Em text
(the file contains only
printing characters and a few common control
characters and is probably safe to read on an
.Dv ASCII
terminal),
.Em executable
(the file contains the result of compiling a program
in a form understandable to some
.Tn UNIX
kernel or another),
or
.Em data
meaning anything else (data is usually
.Dq binary
or non-printable).
Exceptions are well-known file formats (core files, tar archives)
that are known to contain binary data.
When adding local definitions to /etc/magic, make sure to
.Em "preserve these keywords" .
Users depend on knowing that all the readable files in a directory
have the word
.Dq text
printed.
Don't do as Berkeley did and change
.Dq shell commands text
to
.Dq shell script .
.Pp
The filesystem tests are based on examining the return from a
.Xr stat 2
system call.
The program checks to see if the file is empty,
or if it's some sort of special file.
Any known file types appropriate to the system you are running on
(sockets, symbolic links, or named pipes (FIFOs) on those systems that
implement them)
are intuited if they are defined in the system header file
.In sys/stat.h .
.Pp
The magic tests are used to check for files with data in
particular fixed formats.
The canonical example of this is a binary executable (compiled program)
.Dv a.out
file, whose format is defined in
.In elf.h ,
.In a.out.h
and possibly
.In exec.h
in the standard include directory.
These files have a
.Dq "magic number"
stored in a particular place
near the beginning of the file that tells the
.Tn UNIX
operating system
that the file is a binary executable, and which of several types thereof.
The concept of a
.Dq "magic"
has been applied by extension to data files.
Any file with some invariant identifier at a small fixed
offset into the file can usually be described in this way.
The information identifying these files is read from /etc/magic
and the compiled
magic file
.Pa __MAGIC__.mgc ,
or the files in the directory
.Pa __MAGIC__
if the compiled file does not exist.
In addition, if
.Pa $HOME/.magic.mgc
or
.Pa $HOME/.magic
exists, it will be used in preference to the system magic files.
.Pp
If a file does not match any of the entries in the magic file,
it is examined to see if it seems to be a text file.
ASCII, ISO-8859-x, non-ISO 8-bit extended-ASCII character sets
(such as those used on Macintosh and IBM PC systems),
UTF-8-encoded Unicode, UTF-16-encoded Unicode, and EBCDIC
character sets can be distinguished by the different
ranges and sequences of bytes that constitute printable text
in each set.
If a file passes any of these tests, its character set is reported.
ASCII, ISO-8859-x, UTF-8, and extended-ASCII files are identified
as
.Dq text
because they will be mostly readable on nearly any terminal;
UTF-16 and EBCDIC are only
.Dq character data
because, while
they contain text, it is text that will require translation
before it can be read.
In addition,
.Nm
will attempt to determine other characteristics of text-type files.
If the lines of a file are terminated by CR, CRLF, or NEL, instead
of the Unix-standard LF, this will be reported.
Files that contain embedded escape sequences or overstriking
will also be identified.
.Pp
Once
.Nm
has determined the character set used in a text-type file,
it will
attempt to determine in what language the file is written.
The language tests look for particular strings (cf.
.In names.h )
that can appear anywhere in the first few blocks of a file.
For example, the keyword
.Em .br
indicates that the file is most likely a
.Xr troff 1
input file, just as the keyword
.Em struct
indicates a C program.
These tests are less reliable than the previous
two groups, so they are performed last.
The language test routines also test for some miscellany
(such as
.Xr tar 1
archives, JSON files).
.Pp
Any file that cannot be identified as having been written
in any of the character sets listed above is simply said to be
.Dq data .
.Sh OPTIONS
.Bl -tag -width indent
.It Fl Fl apple
Causes the file command to output the file type and creator code as
used by older MacOS versions.
The code consists of eight letters,
the first describing the file type, the latter the creator.
the first describing the file type, the latter the creator.
This option works properly only for file formats that have the
apple-style output defined.
.It Fl b , Fl Fl brief
Do not prepend filenames to output lines (brief mode).
.It Fl C , Fl Fl compile
Write a
.Pa magic.mgc
output file that contains a pre-parsed version of the magic file or directory.
.It Fl c , Fl Fl checking-printout
Cause a checking printout of the parsed form of the magic file.
This is usually used in conjunction with the
.Fl m
flag to debug a new magic file before installing it.
.It Fl d
Prints internal debugging information to stderr.
.It Fl E
On filesystem errors (file not found etc), instead of handling the error
as regular output as POSIX mandates and keep going, issue an error message
and exit.
.It Fl e , Fl Fl exclude Ar testname
Exclude the test named in
.Ar testname
from the list of tests made to determine the file type.
Valid test names are:
.Bl -tag -width compress
.It apptype
.Dv EMX
application type (only on EMX).
.It ascii
Various types of text files (this test will try to guess the text
encoding, irrespective of the setting of the
.Sq encoding
option).
.It encoding
Different text encodings for soft magic tests.
.It tokens
Ignored for backwards compatibility.
.It cdf
Prints details of Compound Document Files.
.It compress
Checks for, and looks inside, compressed files.
.It elf
Prints ELF file details, provided soft magic tests are enabled and the
elf magic is found.
.It json
Examines JSON (RFC-7159) files by parsing them for compliance.
.It soft
Consults magic files.
.It tar
Examines tar files by verifying the checksum of the 512 byte tar header.
Excluding this test can provide more detailed content description by using
the soft magic method.
.It text
A synonym for
.Sq ascii .
.El
.It Fl Fl extension
Print a slash-separated list of valid extensions for the file type found.
.It Fl F , Fl Fl separator Ar separator
Use the specified string as the separator between the filename and the
file result returned.
Defaults to
.Sq \&: .
.It Fl f , Fl Fl files-from Ar namefile
Read the names of the files to be examined from
.Ar namefile
(one per line)
before the argument list.
Either
.Ar namefile
or at least one filename argument must be present;
to test the standard input, use
.Sq -
as a filename argument.
Please note that
.Ar namefile
is unwrapped and the enclosed filenames are processed when this option is
encountered and before any further options processing is done.
This allows one to process multiple lists of files with different command line
arguments on the same
.Nm
invocation.
Thus if you want to set the delimiter, you need to do it before you specify
the list of files, like:
.Dq Fl F Ar @ Fl f Ar namefile ,
instead of:
.Dq Fl f Ar namefile Fl F Ar @ .
.It Fl h , Fl Fl no-dereference
option causes symlinks not to be followed
(on systems that support symbolic links).
This is the default if the environment variable
.Dv POSIXLY_CORRECT
is not defined.
.It Fl i , Fl Fl mime
Causes the file command to output mime type strings rather than the more
traditional human readable ones.
Thus it may say
.Sq text/plain; charset=us-ascii
rather than
.Dq ASCII text .
.It Fl Fl mime-type , Fl Fl mime-encoding
Like
.Fl i ,
but print only the specified element(s).
.It Fl k , Fl Fl keep-going
Don't stop at the first match, keep going.
Subsequent matches will be
have the string
.Sq "\[rs]012\- "
prepended.
(If you want a newline, see the
.Fl r
option.)
The magic pattern with the highest strength (see the
.Fl l
option) comes first.
.It Fl l , Fl Fl list
Shows a list of patterns and their strength sorted descending by
.Xr magic 4
strength
which is used for the matching (see also the
.Fl k
option).
.It Fl L , Fl Fl dereference
option causes symlinks to be followed, as the like-named option in
.Xr ls 1
(on systems that support symbolic links).
This is the default if the environment variable
.Ev POSIXLY_CORRECT
is defined.
.It Fl m , Fl Fl magic-file Ar magicfiles
Specify an alternate list of files and directories containing magic.
This can be a single item, or a colon-separated list.
If a compiled magic file is found alongside a file or directory,
it will be used instead.
.It Fl N , Fl Fl no-pad
Don't pad filenames so that they align in the output.
.It Fl n , Fl Fl no-buffer
Force stdout to be flushed after checking each file.
This is only useful if checking a list of files.
It is intended to be used by programs that want filetype output from a pipe.
.It Fl p , Fl Fl preserve-date
On systems that support
.Xr utime 3
or
.Xr utimes 2 ,
attempt to preserve the access time of files analyzed, to pretend that
.Nm
never read them.
.It Fl P , Fl Fl parameter Ar name=value
Set various parameter limits.

debian/changelog
file (1:5.35.1-1+dde) unstable; urgency=medium

  * Cherry-pick fix for JSON detection. Closes: 922874

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sat, 02 Mar 2019 08:05:54 +0100

file (1:5.35-3) unstable; urgency=medium

  * Cherry-pick many commits since 5.35 release that seem wise to
    include in buster.
    * Closes: #922967 [CVE-2019-8904]
    * Closes: #922968 [CVE-2019-8905 CVE-2019-8907]
    * Closes: #922969 [CVE-2019-8906]
  * Cherry-pick two documentation fix commits

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Fri, 01 Mar 2019 09:27:11 +0100

file (1:5.35-2) unstable; urgency=medium

  * Upload to unstable
  * Clarify manpage: seccomp is disabled for the time being (#917930)

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Fri, 11 Jan 2019 17:26:27 +0100

file (1:5.35-1) experimental; urgency=medium

  * New upstream version 5.35, upload to experimental

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sat, 03 Nov 2018 14:37:11 +0100

file (1:5.34-2) unstable; urgency=medium

  * Cherry-pick from upstream: Handle [elf] files without program
    headers gracefully. Closes: #882310

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sun, 05 Aug 2018 19:53:19 +0200

file (1:5.34-1) experimental; urgency=medium

  * New upstream version 5.34, upload to experimental
  * Fixes:
    - Detect FITS image data. Closes: #857999
    - Don't mis-detect enigmail properties. Closes: #903318
  * Also add preliminary magic for:
    - AutoCAD Drawing Exchange Format. Closes: #702744
    - Extended Display Identification Data (EDID). Closes: #896932
    - SketchUp files. Closes: #903693

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Thu, 26 Jul 2018 07:28:30 +0200

file (1:5.33-3) unstable; urgency=high

  * Avoid reading past the end of buffer. Closes: #901351
    [CVE-2018-10360]

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Mon, 11 Jun 2018 22:52:24 +0200

file (1:5.33-2) unstable; urgency=medium

  * Recognize ia64 and amd64 COFF files. Closes: #877643
  * Disable erroneous DBPF detection that breaks SQLite and some
    more. Closes: 897153

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sun, 29 Apr 2018 12:13:32 +0200

file (1:5.33-1) unstable; urgency=low

  * New upstream version 5.33

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Tue, 24 Apr 2018 07:39:10 +0200

file (1:5.32-2) unstable; urgency=medium

  * No longer ship python-magic libraries, they are now provided
    by src:python-magic

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Tue, 13 Feb 2018 21:43:38 +0100

file (1:5.32-1) unstable; urgency=high

  * New upstream version 5.32

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Mon, 04 Sep 2017 18:20:35 +0200

file (1:5.31-1) unstable; urgency=medium

  * New upstream version 5.31

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sun, 06 Aug 2017 17:29:56 +0200

file (1:5.30-1) unstable; urgency=high

  * New upstream version 5.30
  * Cherry-pick commits that fix issues found by oss-fuzz
  * Revert new features introduced since 1:5.29-3

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Sat, 29 Apr 2017 10:57:11 +0200

file (1:5.29-3) unstable; urgency=medium

  * Restore full local.support-local-definitions-in-etc-magic patch.
    Closes: #852476
  * Include all upstream commits since the 5.29 release
  * Improve detection of Flash data. Closes: #838860

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Thu, 26 Jan 2017 00:29:24 +0100

file (1:5.29-2) unstable; urgency=medium

  * Include all upstream commits since the 5.29 release. Addresses:
    - Detect compiled YARA rules. Closes: #833872
    - Detect old Word for Mac documents. Closes: #842117
  * Disable detection of Algol68 files, way too many false positives

 -- Christoph Biedl <debian.axhn@manchmal.in-ulm.de>  Tue, 13 Dec 2016 16:06:43 +0100

file (1:5.29-1) unstable; urgency=medium

  * New upstream version 5.29. Addresses (in order of appearance):
    - "no read permission" if read from empty pipe. Closes: #508963
    - Mistakes hungarian text as dos executables. Closes: #641012
    - Does not recognize some GPG key public rings. Closes: #729286
    - Show more information for MySQL files. Closes: #751826
    - Linux kernel version string truncated. Closes: #756949
    - Document file's '-d' option. Closes: #764462
    - Detect JPEG-XR. Closes: #771303
    - Detect Material exchange container format (mxf). Closes: #782744
    - Strengthen detection of Embedded OpenType (EOT). Closes: #784572
    - Mistakes some text as bitmap. Closes: #799352
    - Dectect swp files from nano, vim, and kate. Closes: #803219
    - Mistakes some SVG files as HTML. Closes: #829199
  * Fix FTCBFS: Remove stage1 profile in favour of a proper arch/indep
    split. Thanks Helmut Grohne. Closes: #841030


src/file.c
/*
 * Copyright (c) Ian F. Darwin 1986-1995.
 * Software written by Ian F. Darwin and others;
 * maintained 1995-present by Christos Zoulas and others.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice immediately at the beginning of the file, without modification,
 *    this list of conditions, and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/*
 * file - find type of a file or files - main program.
 */

#include "file.h"

#ifndef	lint
FILE_RCSID("@(#)$File: file.c,v 1.178 2018/10/01 18:50:31 christos Exp $")
#endif	/* lint */

#include "magic.h"

#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#ifdef RESTORE_TIME
# if (__COHERENT__ >= 0x420)
#  include <sys/utime.h>
# else
#  ifdef USE_UTIMES
#   include <sys/time.h>
#  else
#   include <utime.h>
#  endif
# endif
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>	/* for read() */
#endif
#ifdef HAVE_WCHAR_H
#include <wchar.h>
#endif

#if defined(HAVE_GETOPT_H) && defined(HAVE_STRUCT_OPTION)
# include <getopt.h>
# ifndef HAVE_GETOPT_LONG
int getopt_long(int, char * const *, const char *,
    const struct option *, int *);
# endif
# else
#  include "mygetopt.h"
#endif

#ifdef S_IFLNK
# define IFLNK_h "h"
# define IFLNK_L "L"
#else
# define IFLNK_h ""
# define IFLNK_L ""
#endif

#ifdef HAVE_LIBSECCOMP
# define SECCOMP_S "S"
#else
# define SECCOMP_S ""
#endif

#define FILE_FLAGS	"bcCdE" IFLNK_h "ik" IFLNK_L "lNnprs" SECCOMP_S "vzZ0"
#define OPTSTRING	"bcCde:Ef:F:hiklLm:nNpP:rsSvzZ0"

# define USAGE  \
    "Usage: %s [-" FILE_FLAGS "] [--apple] [--extension] [--mime-encoding]\n" \
    "            [--mime-type] [-e <testname>] [-F <separator>] " \
    " [-f <namefile>]\n" \
    "            [-m <magicfiles>] [-P <parameter=value>] <file> ...\n" \
    "       %s -C [-m <magicfiles>]\n" \
    "       %s [--help]\n"

private int 		/* Global command-line options 		*/
	bflag = 0,	/* brief output format	 		*/
	nopad = 0,	/* Don't pad output			*/
	nobuffer = 0,   /* Do not buffer stdout 		*/
	nulsep = 0;	/* Append '\0' to the separator		*/

private const char *separator = ":";	/* Default field separator	*/
private const struct option long_options[] = {
#define OPT_HELP		1
#define OPT_APPLE		2
#define OPT_EXTENSIONS		3
#define OPT_MIME_TYPE		4
#define OPT_MIME_ENCODING	5
#define OPT(shortname, longname, opt, def, doc)      \
    {longname, opt, NULL, shortname},
#define OPT_LONGONLY(longname, opt, def, doc, id)        \
    {longname, opt, NULL, id},
#include "file_opts.h"
#undef OPT
#undef OPT_LONGONLY
    {0, 0, NULL, 0}
    };

private const struct {
	const char *name;
	int value;
} nv[] = {
	{ "apptype",	MAGIC_NO_CHECK_APPTYPE },
	{ "ascii",	MAGIC_NO_CHECK_ASCII },
	{ "cdf",	MAGIC_NO_CHECK_CDF },
	{ "compress",	MAGIC_NO_CHECK_COMPRESS },
	{ "elf",	MAGIC_NO_CHECK_ELF },
	{ "encoding",	MAGIC_NO_CHECK_ENCODING },
	{ "soft",	MAGIC_NO_CHECK_SOFT },
	{ "tar",	MAGIC_NO_CHECK_TAR },
	{ "json",	MAGIC_NO_CHECK_JSON },
	{ "text",	MAGIC_NO_CHECK_TEXT },	/* synonym for ascii */
	{ "tokens",	MAGIC_NO_CHECK_TOKENS }, /* OBSOLETE: ignored for backwards compatibility */
};

private struct {
	const char *name;
	int tag;
	size_t value;
	int set;
} pm[] = {
	{ "indir",	MAGIC_PARAM_INDIR_MAX, 0, 0 },
	{ "name",	MAGIC_PARAM_NAME_MAX, 0, 0 },
	{ "elf_phnum",	MAGIC_PARAM_ELF_PHNUM_MAX, 0, 0 },
	{ "elf_shnum",	MAGIC_PARAM_ELF_SHNUM_MAX, 0, 0 },
	{ "elf_notes",	MAGIC_PARAM_ELF_NOTES_MAX, 0, 0 },
	{ "regex",	MAGIC_PARAM_REGEX_MAX, 0, 0 },
	{ "bytes",	MAGIC_PARAM_BYTES_MAX, 0, 0 },
};

private int posixly;

#ifdef __dead
__dead
#endif
private void usage(void);
private void docprint(const char *, int);
#ifdef __dead
__dead
#endif
private void help(void);

private int unwrap(struct magic_set *, const char *);
private int process(struct magic_set *ms, const char *, int);
private struct magic_set *load(const char *, int);
private void setparam(const char *);
private void applyparam(magic_t);


/*
 * main - parse arguments and handle options
 */
int
main(int argc, char *argv[])
{
	int c;
	size_t i;
	int action = 0, didsomefiles = 0, errflg = 0;
	int flags = 0, e = 0;
#ifdef HAVE_LIBSECCOMP
	int sandbox = 1;
#endif
	struct magic_set *magic = NULL;
	int longindex;
	const char *magicfile = NULL;		/* where the magic is	*/
	char *progname;

	/* makes islower etc work for other langs */
#ifdef HAVE_SETLOCALE
	(void)setlocale(LC_CTYPE, "");
#endif

#ifdef __EMX__
	/* sh-like wildcard expansion! Shouldn't hurt at least ... */
	_wildcard(&argc, &argv);
#endif

	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	file_setprogname(progname);


#ifdef S_IFLNK
	posixly = getenv("POSIXLY_CORRECT") != NULL;
	flags |=  posixly ? MAGIC_SYMLINK : 0;
#endif
	while ((c = getopt_long(argc, argv, OPTSTRING, long_options,
	    &longindex)) != -1)
		switch (c) {
		case OPT_HELP:
			help();
			break;
		case OPT_APPLE:
			flags |= MAGIC_APPLE;
			break;
		case OPT_EXTENSIONS:
			flags |= MAGIC_EXTENSION;
			break;
		case OPT_MIME_TYPE:
			flags |= MAGIC_MIME_TYPE;
			break;
		case OPT_MIME_ENCODING:
			flags |= MAGIC_MIME_ENCODING;
			break;
		case '0':
			nulsep++;
			break;
		case 'b':
			bflag++;
			break;
		case 'c':
			action = FILE_CHECK;
			break;
		case 'C':
			action = FILE_COMPILE;
			break;
		case 'd':
			flags |= MAGIC_DEBUG|MAGIC_CHECK;
			break;
		case 'E':
			flags |= MAGIC_ERROR;
			break;
		case 'e':
			for (i = 0; i < sizeof(nv) / sizeof(nv[0]); i++)
				if (strcmp(nv[i].name, optarg) == 0)
					break;

			if (i == sizeof(nv) / sizeof(nv[0]))
				errflg++;
			else
				flags |= nv[i].value;
			break;

		case 'f':
			if(action)
				usage();
			if (magic == NULL)
				if ((magic = load(magicfile, flags)) == NULL)
					return 1;
			applyparam(magic);
			e |= unwrap(magic, optarg);
			++didsomefiles;
			break;
		case 'F':
			separator = optarg;
			break;
		case 'i':
			flags |= MAGIC_MIME;
			break;
		case 'k':
			flags |= MAGIC_CONTINUE;
			break;
		case 'l':
			action = FILE_LIST;
			break;
		case 'm':
			magicfile = optarg;
			break;
		case 'n':
			++nobuffer;
			break;
		case 'N':
			++nopad;
			break;
#if defined(HAVE_UTIME) || defined(HAVE_UTIMES)
		case 'p':
			flags |= MAGIC_PRESERVE_ATIME;
			break;
#endif
		case 'P':
			setparam(optarg);
			break;
		case 'r':
			flags |= MAGIC_RAW;
			break;
		case 's':
			flags |= MAGIC_DEVICES;
			break;
#ifdef HAVE_LIBSECCOMP
		case 'S':
			sandbox = 0;
			break;
#endif
		case 'v':
			if (magicfile == NULL)
				magicfile = magic_getpath(magicfile, action);
			(void)fprintf(stdout, "%s-%s\n", file_getprogname(),
			    VERSION);
			(void)fprintf(stdout, "magic file from %s\n",
			    magicfile);
			return 0;
		case 'z':
			flags |= MAGIC_COMPRESS;
			break;

		case 'Z':
			flags |= MAGIC_COMPRESS|MAGIC_COMPRESS_TRANSP;
			break;
#ifdef S_IFLNK
		case 'L':
			flags |= MAGIC_SYMLINK;
			break;
		case 'h':
			flags &= ~MAGIC_SYMLINK;
			break;
#endif
		case '?':
		default:
			errflg++;
			break;
		}

	if (errflg) {
		usage();
	}
	if (e)
		return e;

#ifdef HAVE_LIBSECCOMP
#if 0
	if (sandbox && enable_sandbox_basic() == -1)
#else
	if (sandbox && enable_sandbox_full() == -1)
#endif
		file_err(EXIT_FAILURE, "SECCOMP initialisation failed");
#endif /* HAVE_LIBSECCOMP */

	if (MAGIC_VERSION != magic_version())
		file_warnx("Compiled magic version [%d] "
		    "does not match with shared library magic version [%d]\n",
		    MAGIC_VERSION, magic_version());

	switch(action) {
	case FILE_CHECK:
	case FILE_COMPILE:
	case FILE_LIST:
		/*
		 * Don't try to check/compile ~/.magic unless we explicitly
		 * ask for it.
		 */
		magic = magic_open(flags|MAGIC_CHECK);
		if (magic == NULL) {
			file_warn("Can't create magic");
			return 1;
		}


		switch(action) {
		case FILE_CHECK:
			c = magic_check(magic, magicfile);
			break;
		case FILE_COMPILE:
			c = magic_compile(magic, magicfile);
			break;
		case FILE_LIST:
			c = magic_list(magic, magicfile);
			break;
		default:
			abort();
		}
		if (c == -1) {
			file_warnx("%s", magic_error(magic));
			e = 1;
			goto out;
		}
		goto out;
	default:
		if (magic == NULL)
			if ((magic = load(magicfile, flags)) == NULL)
				return 1;
		applyparam(magic);
	}

	if (optind == argc) {
		if (!didsomefiles)
			usage();
	}
	else {
		size_t j, wid, nw;
		for (wid = 0, j = CAST(size_t, optind); j < CAST(size_t, argc);
		    j++) {
			nw = file_mbswidth(argv[j]);
			if (nw > wid)
				wid = nw;
		}
		/*
		 * If bflag is only set twice, set it depending on
		 * number of files [this is undocumented, and subject to change]
		 */
		if (bflag == 2) {
			bflag = optind >= argc - 1;
		}
		for (; optind < argc; optind++)
			e |= process(magic, argv[optind], wid);
	}

out:
	if (magic)
		magic_close(magic);
	return e;
}

private void
applyparam(magic_t magic)
{
	size_t i;

	for (i = 0; i < __arraycount(pm); i++) {
		if (!pm[i].set)
			continue;
		if (magic_setparam(magic, pm[i].tag, &pm[i].value) == -1)
			file_err(EXIT_FAILURE, "Can't set %s", pm[i].name);
	}
}

private void
setparam(const char *p)
{
	size_t i;
	char *s;

	if ((s = strchr(p, '=')) == NULL)
		goto badparm;

	for (i = 0; i < __arraycount(pm); i++) {
		if (strncmp(p, pm[i].name, s - p) != 0)
			continue;
		pm[i].value = atoi(s + 1);
		pm[i].set = 1;
		return;
	}
badparm:
	file_errx(EXIT_FAILURE, "Unknown param %s", p);
}

private struct magic_set *
/*ARGSUSED*/
load(const char *magicfile, int flags)
{
	struct magic_set *magic = magic_open(flags);
	const char *e;

	if (magic == NULL) {
		file_warn("Can't create magic");
		return NULL;
	}
	if (magic_load(magic, magicfile) == -1) {
		file_warn("%s", magic_error(magic));
		magic_close(magic);
		return NULL;
	}
	if ((e = magic_error(magic)) != NULL)
		file_warn("%s", e);
	return magic;
}

/*
 * unwrap -- read a file of filenames, do each one.
 */
private int
unwrap(struct magic_set *ms, const char *fn)
{
	FILE *f;
	ssize_t len;
	char *line = NULL;
	size_t llen = 0;
	int wid = 0, cwid;
	int e = 0;

	if (strcmp("-", fn) == 0) {
		f = stdin;
		wid = 1;
	} else {
		if ((f = fopen(fn, "r")) == NULL) {
			file_warn("Cannot open `%s'", fn);
			return 1;
		}

		while ((len = getline(&line, &llen, f)) > 0) {
			if (line[len - 1] == '\n')
				line[len - 1] = '\0';
			cwid = file_mbswidth(line);
			if (cwid > wid)
				wid = cwid;
		}

		rewind(f);
	}

	while ((len = getline(&line, &llen, f)) > 0) {
		if (line[len - 1] == '\n')
			line[len - 1] = '\0';
		e |= process(ms, line, wid);
		if(nobuffer)
			(void)fflush(stdout);
	}

	free(line);
	(void)fclose(f);
	return e;
}

/*
 * Called for each input file on the command line (or in a list of files)
 */
private int
process(struct magic_set *ms, const char *inname, int wid)
{
	const char *type, c = nulsep > 1 ? '\0' : '\n';
	int std_in = strcmp(inname, "-") == 0;

	if (wid > 0 && !bflag) {
		(void)printf("%s", std_in ? "/dev/stdin" : inname);
		if (nulsep)
			(void)putc('\0', stdout);
		if (nulsep < 2) {
			(void)printf("%s", separator);
			(void)printf("%*s ", CAST(int, nopad ? 0
			    : (wid - file_mbswidth(inname))), "");
		}
	}

	type = magic_file(ms, std_in ? NULL : inname);

	if (type == NULL) {
		(void)printf("ERROR: %s%c", magic_error(ms), c);
		return 1;
	} else {
		(void)printf("%s%c", type, c);
		return 0;
	}
}

protected size_t
file_mbswidth(const char *s)
{
#if defined(HAVE_WCHAR_H) && defined(HAVE_MBRTOWC) && defined(HAVE_WCWIDTH)
	size_t bytesconsumed, old_n, n, width = 0;
	mbstate_t state;
	wchar_t nextchar;
	(void)memset(&state, 0, sizeof(mbstate_t));
	old_n = n = strlen(s);

	while (n > 0) {
		bytesconsumed = mbrtowc(&nextchar, s, n, &state);
		if (bytesconsumed == CAST(size_t, -1) ||
		    bytesconsumed == CAST(size_t, -2)) {
			/* Something went wrong, return something reasonable */
			return old_n;
		}
		if (s[0] == '\n') {
			/*
			 * do what strlen() would do, so that caller
			 * is always right
			 */
			width++;
		} else {
			int w = wcwidth(nextchar);
			if (w > 0)
				width += w;
		}

		s += bytesconsumed, n -= bytesconsumed;
	}
	return width;
#else
	return strlen(s);
#endif
}

private void
usage(void)
{
	const char *pn = file_getprogname();
	(void)fprintf(stderr, USAGE, pn, pn, pn);
	exit(EXIT_FAILURE);
}

private void
defprint(int def)
{
	if (!def)
		return;
	if (((def & 1) && posixly) || ((def & 2) && !posixly))
		fprintf(stdout, " (default)");
	fputc('\n', stdout);
}

private void
docprint(const char *opts, int def)
{
	size_t i;
	int comma;
	char *sp, *p;

	p = strstr(opts, "%o");
	if (p == NULL) {
		fprintf(stdout, "%s", opts);
		defprint(def);
		return;
	}

	for (sp = p - 1; sp > opts && *sp == ' '; sp--)
		continue;

	fprintf(stdout, "%.*s", CAST(int, p - opts), opts);

	comma = 0;
	for (i = 0; i < __arraycount(nv); i++) {
		fprintf(stdout, "%s%s", comma++ ? ", " : "", nv[i].name);
		if (i && i % 5 == 0 && i != __arraycount(nv) - 1) {
			fprintf(stdout, ",\n%*s", CAST(int, p - sp - 1), "");
			comma = 0;
		}
	}

	fprintf(stdout, "%s", opts + (p - opts) + 2);
}

private void
help(void)
{
	(void)fputs(
"Usage: file [OPTION...] [FILE...]\n"
"Determine type of FILEs.\n"
"\n", stdout);
#define OPT(shortname, longname, opt, def, doc)      \
	fprintf(stdout, "  -%c, --" longname, shortname), \
	docprint(doc, def);
#define OPT_LONGONLY(longname, opt, def, doc, id)        \
	fprintf(stdout, "      --" longname),	\
	docprint(doc, def);
#include "file_opts.h"
#undef OPT
#undef OPT_LONGONLY
	fprintf(stdout, "\nReport bugs to https://bugs.astron.com/\n");
	exit(EXIT_SUCCESS);
}

private const char *file_progname;

protected void
file_setprogname(const char *progname)
{
	file_progname = progname;
}

protected const char *
file_getprogname(void)
{
	return file_progname;
}

protected void
file_err(int e, const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	fprintf(stderr, "%s: ", file_progname);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, " (%s)\n", strerror(se));
	exit(e);
}

protected void
file_errx(int e, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	fprintf(stderr, "%s: ", file_progname);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n");
	exit(e);
}

protected void
file_warn(const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	fprintf(stderr, "%s: ", file_progname);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, " (%s)\n", strerror(se));
	errno = se;
}

protected void
file_warnx(const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	fprintf(stderr, "%s: ", file_progname);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n");
	errno = se;
}
